### PR TITLE
 Rename `UninhabitedEnumBranching` to `UnreachableEnumBranching` 

### PR DESCRIPTION
**Pull request by dianqk**
Per [#120268](https://github.com/rust-lang/rust/pull/120268#discussion_r1517492060), I rename `UninhabitedEnumBranching` to `UnreachableEnumBranching` .

I solved some nits to add some comments.

I adjusted the workaround restrictions. This should be useful for `a <= b` and `if let Some/Ok(v)`. For enum with few variants, `early-tailduplication` should not cause compile time overhead.

r? RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
Sorry, I can't review any behavior change in a MIR opt.
r? compiler

**Comment by DianQK** [comment]
@cjgillot Would you be willing to add yourself to `mir-opt` in `triagebot.toml`. Then I can use `r? mir-opt`. :)

File: compiler/rustc_mir_transform/src/unreachable_enum_branching.rs [hunk 1]
```diff
@@ -136,11 +155,32 @@ impl<'tcx> MirPass<'tcx> for UninhabitedEnumBranching {
...
+                // Despite the LLVM issue, we hope that small enum can still be transformed.
+                // This is valuable for both `a <= b` and `if let Some/Ok(v)`.
+                && (targets.all_targets().len() <= 3
+                    || check_successors(&body.basic_blocks, targets.otherwise()));
             let replace_otherwise_to_unreachable = otherwise_is_last_variant
-                || !otherwise_is_empty_unreachable && allowed_variants.is_empty();
+                || (!otherwise_is_empty_unreachable && allowed_variants.is_empty());
 
             if unreachable_targets.is_empty() && !replace_otherwise_to_unreachable {
                 continue;
```
Comment (riking) [inline compiler/rustc_mir_transform/src/unreachable_enum_branching.rs:186]: The `allow(rustc::potential_query_instability)` should include a comment that the length of the list is known to be 1.

**Pull request merged**